File: void-js/src/engine/grid/grid.test.ts
import {describe, test} from 'node:test'
import {assert} from '../../test/assert.ts'
import type {Void} from '../void.ts'
import {Grid, type GridEnt, type GridOpts} from './grid.ts'

const opts: Readonly<GridOpts> = {
  cellSize: 10,
  xy: {x: -45, y: -45},
  wh: {w: 100, h: 100}
}

describe('insert()', () => {
  test('empty grid yields no pairs', () => {
    const grid = new Grid(opts)
    assert(pairs(grid), [])
  })

  test('single ent yields no pairs', () => {
    const grid = new Grid(opts)
    grid.insert(Ent(5, 5))
    assert(pairs(grid), [])
  })

  test('out-of-bounds ents are dropped', () => {
    const grid = new Grid(opts)
    grid.insert(Ent(200, 200))
    grid.insert(Ent(200, 200))
    grid.insert(Ent(200, 200))
    grid.insert(Ent(200, 200))
    assert(pairs(grid), [])
  })

  test('negative-coord ents are dropped', () => {
    const grid = new Grid(opts)
    grid.insert(Ent(-1, -1))
    assert(pairs(grid), [])
  })
})

describe('forEachCollision()', () => {
  describe('same cell', () => {
    test('two ents produce one pair', () => {
      const grid = new Grid(opts)
      const a = Ent(5, 5)
      const b = Ent(6, 6)
      grid.insert(a)
      grid.insert(b)
      assert(pairs(grid), [[a, b]])
    })

    test('three ents produce three pairs', () => {
      const grid = new Grid(opts)
      const a = Ent(5, 5)
      const b = Ent(6, 6)
      const c = Ent(7, 7)
      grid.insert(a)
      grid.insert(b)
      grid.insert(c)
      assert(pairs(grid), [
        [a, b],
        [a, c],
        [b, c]
      ])
    })
  })

  describe('adjacent cells', () => {
    for (const [name, bx, by] of [
      ['(4, 4) above-left', -5, -5],
      ['(4, 5) above', 5, -5],
      ['(4, 6) above-right', 15, -5],
      ['(5, 4) left', -5, 5],
      ['(5, 5) same', 5, 5],
      ['(5, 6) right', 15, 5],
      ['(6, 4) below-left', -5, 15],
      ['(6, 5) below', 5, 15],
      ['(6, 6) below-right', 15, 15]
    ] as const) {
      test(`A in cell (5, 5), B in cell ${name}`, () => {
        const grid = new Grid(opts)
        const a = Ent(5, 5) // cell (5, 5).
        const b = Ent(bx, by)
        grid.insert(a)
        grid.insert(b)
        assert(pairs(grid).length, 1)
      })
    }
  })

  describe('no spurious pairs', () => {
    test('ents two cells apart produce no pairs', () => {
      const grid = new Grid(opts)
      grid.insert(Ent(5, 5)) // cell (5, 5).
      grid.insert(Ent(25, 5)) // cell (5, 7).
      assert(pairs(grid), [])
    })

    test('16 border cells of 5×5 neighborhood produce no pairs with A', () => {
      // A in cell (5, 5). border cells are those at row or col 3 or 7, two
      // cells away, and should never pair with A.
      //                            row
      //      ┌───┬───┬───┬───┬───┐
      //      │ · │ · │ · │ · │ · │ 3
      //      ├───┼───┼───┼───┼───┤
      //      │ · │   │   │   │ · │ 4
      //      ├───┼───┼───┼───┼───┤
      //      │ · │   │ A │   │ · │ 5
      //      ├───┼───┼───┼───┼───┤
      //      │ · │   │   │   │ · │ 6
      //      ├───┼───┼───┼───┼───┤
      //      │ · │ · │ · │ · │ · │ 7
      //      └───┴───┴───┴───┴───┘
      //  col   3   4   5   6   7
      const grid = new Grid(opts)
      const borderXYs: [number, number][] = [
        [-15, -15],
        [-5, -15],
        [5, -15],
        [15, -15],
        [25, -15], // row 3
        [-15, -5],
        [-15, 5],
        [-15, 15], // col 3
        [25, -5],
        [25, 5],
        [25, 15], // col 7
        [-15, 25],
        [-5, 25],
        [5, 25],
        [15, 25],
        [25, 25] // row 7
      ]
      assert(borderXYs.length, 16)
      for (const [x, y] of borderXYs) grid.insert(Ent(x, y))
      const a = Ent(5, 5) // cell (5, 5).
      grid.insert(a)
      const result = pairs(grid)
      assert(
        result.every(([hitA, hitB]) => hitA !== a && hitB !== a),
        true
      )
    })

    test('each pair reported exactly once', () => {
      const grid = new Grid(opts)
      const a = Ent(5, 5)
      const b = Ent(14, 5) // cell (5, 6).
      const c = Ent(15, 5) // cell (5, 6).
      grid.insert(a)
      grid.insert(b)
      grid.insert(c)
      const result = pairs(grid)
      assert(result.length, 3)
    })

    test('nonzero origin shifts bucketing correctly', () => {
      const grid = new Grid({
        xy: {x: 50, y: 50},
        wh: {w: 100, h: 100},
        cellSize: 10
      })
      const a = Ent(55, 55) // cell (1, 1).
      const b = Ent(65, 55) // cell (1, 2). adjacent.
      const c = Ent(80, 55) // cell (1, 3). two away from a.
      grid.insert(a)
      grid.insert(b)
      grid.insert(c)
      const result = pairs(grid)
      // a-b adjacent, b-c adjacent, a-c two apart
      assert(result.length, 2)
    })

    test('4x4 grid one ent per cell, each pair reported exactly once', () => {
      // cell layout (row-major):
      //  0  1  2  3
      //  4  5  6  7
      //  8  9  a  b
      //  c  d  e  f
      // scenario: when processing cell 5 we check 5 vs 6, 8, 9, a (forward
      // neighbors). when we later process cell 9 its forward neighbors are a,
      // c, d, e; not 5. so (5, 9) must appear exactly once.
      const grid = new Grid({
        xy: {x: 0, y: 0},
        wh: {w: 40, h: 40},
        cellSize: 10
      })
      const ents = Array.from({length: 16}, (_, i) => {
        const col = i % 4
        const row = Math.floor(i / 4)
        return Ent(col * 10, row * 10)
      })
      for (const e of ents) grid.insert(e)

      const result = pairs(grid)

      // no pair (a, b) or (b, a) should appear more than once.
      const seen = new Set<string>()
      for (const [a, b] of result) {
        const ia = ents.indexOf(a)
        const ib = ents.indexOf(b)
        const key = `${Math.min(ia, ib)},${Math.max(ia, ib)}`
        assert(seen.has(key), false)
        seen.add(key)
      }

      // (5, 9) appears exactly once.
      const e5 = ents[5]!
      const e9 = ents[9]!
      const count59 = result.filter(
        ([a, b]) => (a === e5 && b === e9) || (a === e9 && b === e5)
      ).length
      assert(count59, 1)

      // 12 right + 12 below + 9 below-right + 9 below-left = 42 total pairs.
      assert(result.length, 42)
    })
  })
})

function Ent(x: number, y: number) {
  return {sprite: {x, y}}
}

function pairs(grid: Readonly<Grid<GridEnt>>): [GridEnt, GridEnt][] {
  const pairs: [GridEnt, GridEnt][] = []
  grid.forEachCollision((a, b) => pairs.push([a, b]), {} as Void)
  return pairs
}

File: src/void/vgrid/grid_test.go
package vgrid

import (
	"testing"

	"github.com/oidoid/void/src/void/vgeo"
)

var testBounds = vgeo.NewBox(
	float32(-45), float32(-45), float32(55), float32(55),
)

func TestGridInsert(t *testing.T) {
	tests := []struct {
		name string
		vals []vgeo.XY[float32]
	}{
		{name: "empty grid yields no pairs"},
		{name: "single ent yields no pairs", vals: []vgeo.XY[float32]{xy(5, 5)}},
		{
			name: "out-of-bounds ents are dropped",
			vals: []vgeo.XY[float32]{xy(200, 200), xy(200, 200), xy(200, 200)},
		},
		{
			name: "negative-coordinate ents are in bounds",
			vals: []vgeo.XY[float32]{xy(-1, -1)},
		},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			grid := newGrid()
			for i, val := range test.vals {
				grid.InsertAt(val, int32(i))
			}
			if got := pairs(&grid); len(got) != 0 {
				t.Errorf("pairs = %v, want none", got)
			}
		})
	}
}

func TestGridSameCell(t *testing.T) {
	tests := []struct {
		name string
		vals []vgeo.XY[float32]
		want int
	}{
		{
			name: "two ents produce one pair",
			vals: []vgeo.XY[float32]{xy(5, 5), xy(6, 6)},
			want: 1,
		},
		{
			name: "three ents produce three pairs",
			vals: []vgeo.XY[float32]{xy(5, 5), xy(6, 6), xy(7, 7)},
			want: 3,
		},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			grid := newGrid()
			for i, val := range test.vals {
				grid.InsertAt(val, int32(i))
			}
			if got := len(pairs(&grid)); got != test.want {
				t.Errorf("pair count = %d, want %d", got, test.want)
			}
		})
	}
}

// a ball is a point at the top-left corner of a box no larger than a cell,
// so only same, right, down, and down-right neighbors can ever truly
// overlap; above-right and below-left never can, regardless of exact
// position within the cell.
func TestGridAdjacentCells(t *testing.T) {
	tests := []struct {
		name string
		xy   vgeo.XY[float32]
		want int
	}{
		{name: "above-left", xy: xy(-5, -5), want: 1},
		{name: "above", xy: xy(5, -5), want: 1},
		{name: "above-right", xy: xy(15, -5), want: 0},
		{name: "left", xy: xy(-5, 5), want: 1},
		{name: "same", xy: xy(5, 5), want: 1},
		{name: "right", xy: xy(15, 5), want: 1},
		{name: "below-left", xy: xy(-5, 15), want: 0},
		{name: "below", xy: xy(5, 15), want: 1},
		{name: "below-right", xy: xy(15, 15), want: 1},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			grid := newGrid()
			grid.InsertAt(xy(5, 5), 0)
			grid.InsertAt(test.xy, 1)
			if got := len(pairs(&grid)); got != test.want {
				t.Errorf("pair count = %d, want %d", got, test.want)
			}
		})
	}
}

func TestGridNoSpuriousPairs(t *testing.T) {
	t.Run("ents two cells apart produce no pairs", func(t *testing.T) {
		grid := newGrid()
		grid.InsertAt(xy(5, 5), 0)  // cell (5, 5).
		grid.InsertAt(xy(25, 5), 1) // cell (5, 7).
		if got := len(pairs(&grid)); got != 0 {
			t.Errorf("pair count = %d, want 0", got)
		}
	})
	t.Run("edge cells produce no pairs with center", func(t *testing.T) {
		// A in cell (5, 5). edge cells are those at row or col 3 or 7, two
		// cells away, and should never pair with A.
		//                            row
		//      ┌───┬───┬───┬───┬───┐
		//      │ · │ · │ · │ · │ · │ 3
		//      ├───┼───┼───┼───┼───┤
		//      │ · │   │   │   │ · │ 4
		//      ├───┼───┼───┼───┼───┤
		//      │ · │   │ A │   │ · │ 5
		//      ├───┼───┼───┼───┼───┤
		//      │ · │   │   │   │ · │ 6
		//      ├───┼───┼───┼───┼───┤
		//      │ · │ · │ · │ · │ · │ 7
		//      └───┴───┴───┴───┴───┘
		//  col   3   4   5   6   7
		grid := newGrid()
		vals := []vgeo.XY[float32]{
			xy(-15, -15), xy(-5, -15), xy(5, -15), xy(15, -15), xy(25, -15), // row 3
			xy(-15, -5), xy(-15, 5), xy(-15, 15), // col 3
			xy(25, -5), xy(25, 5), xy(25, 15), // col 7
			xy(-15, 25), xy(-5, 25), xy(5, 25), xy(15, 25),
			xy(25, 25), // row 7
		}
		for i, val := range vals {
			grid.InsertAt(val, int32(i))
		}
		const center = int32(16)
		grid.InsertAt(xy(5, 5), center) // cell (5, 5).
		for _, pair := range pairs(&grid) {
			if pair[0] == center || pair[1] == center {
				t.Errorf("unexpected center pair %v", pair)
			}
		}
	})
	t.Run("each pair reported exactly once", func(t *testing.T) {
		grid := newGrid()
		vals := []vgeo.XY[float32]{
			xy(5, 5),  // cell (5, 5).
			xy(14, 5), // cell (5, 5). same cell: floor((14+45)/10) = 5.
			xy(15, 5), // cell (6, 5). right neighbor of the first two.
		}
		for i, val := range vals {
			grid.InsertAt(val, int32(i))
		}
		assertUniquePairs(t, pairs(&grid), 3)
	})
	t.Run("nonzero origin shifts bucketing correctly", func(t *testing.T) {
		grid := New(vgeo.NewBox(
			float32(50), float32(50), float32(150), float32(150),
		), 10, 16)
		vals := []vgeo.XY[float32]{
			xy(55, 55), // cell (0, 0).
			xy(65, 55), // cell (1, 0). adjacent, right of the first.
			xy(80, 55), // cell (3, 0). two away from the second.
		}
		for i, val := range vals {
			grid.InsertAt(val, int32(i))
		}
		if got := len(pairs(&grid)); got != 1 {
			t.Errorf("pair count = %d, want 1", got)
		}
	})
	t.Run("4 by 4 grid reports 33 unique pairs", func(t *testing.T) {
		grid := New(vgeo.NewBox(
			float32(0), float32(0), float32(40), float32(40),
		), 10, 16)
		// cell layout (row-major):
		//  0  1  2  3
		//  4  5  6  7
		//  8  9  a  b
		//  c  d  e  f
		// scenario: when processing cell 5 we check 5 vs 6, 9, a (forward
		// neighbors: right, down, down-right). when we later process cell 6
		// its forward neighbors are 7, a, b; not 5. so (5, 6) must appear
		// exactly once.
		for i := range 16 {
			grid.InsertAt(xy(float32(i%4*10), float32(i/4*10)), int32(i))
		}
		// 12 right + 12 down + 9 down-right = 33 total pairs.
		assertUniquePairs(t, pairs(&grid), 33)
	})
}

// balls piling up at (or near) one spot must still report every pair exactly
// once; this is the case that most stresses a spatial grid.
func TestGridDenseCell(t *testing.T) {
	const ballCount = 200
	grid := newGrid()
	for i := range ballCount {
		grid.InsertAt(xy(5, 5), int32(i))
	}
	want := ballCount * (ballCount - 1) / 2
	assertUniquePairs(t, pairs(&grid), want)
}

func newGrid() Grid { return New(testBounds, 10, 16) }

func pairs(grid *Grid) [][2]int32 {
	vals := make([][2]int32, 0)
	grid.ForEach(func(a, b int32) bool {
		vals = append(vals, [2]int32{a, b})
		return false
	})
	return vals
}

func assertUniquePairs(t *testing.T, pairs [][2]int32, want int) {
	t.Helper()
	seen := make(map[[2]int32]bool, len(pairs))
	for _, pair := range pairs {
		if pair[0] > pair[1] {
			pair[0], pair[1] = pair[1], pair[0]
		}
		if seen[pair] {
			t.Errorf("duplicate pair %v", pair)
		}
		seen[pair] = true
	}
	if got := len(pairs); got != want {
		t.Errorf("pair count = %d, want %d", got, want)
	}
}

func xy(x, y float32) vgeo.XY[float32] {
	return vgeo.NewXY(x, y)
}

// reporting a pair resolved must drop both vals from all later pairs, even
// ones not yet reached in the same cell's chain.
func TestGridResolvedPairsAreSkipped(t *testing.T) {
	grid := newGrid()
	for i := range 4 {
		grid.InsertAt(xy(5, 5), int32(i))
	}
	var seen [][2]int32
	grid.ForEach(func(l, r int32) bool {
		seen = append(seen, [2]int32{l, r})
		return len(seen) == 1 // resolve only the first pair encountered.
	})
	if len(seen) != 2 {
		t.Fatalf("calls = %d, want 2", len(seen))
	}
	resolved := map[int32]bool{seen[0][0]: true, seen[0][1]: true}
	if resolved[seen[1][0]] || resolved[seen[1][1]] {
		t.Errorf("pair %v reuses a resolved val from %v", seen[1], seen[0])
	}
}

// resolving a same-cell pair must also prevent those vals from reaching
// cross-cell pairs later in the same pass.
func TestGridResolvedPairsSkipCrossCellPairs(t *testing.T) {
	grid := newGrid()
	grid.InsertAt(xy(5, 5), 0)
	grid.InsertAt(xy(6, 6), 1)  // same cell as val 0.
	grid.InsertAt(xy(15, 5), 2) // right neighbor cell.
	var seen [][2]int32
	grid.ForEach(func(l, r int32) bool {
		seen = append(seen, [2]int32{l, r})
		return true
	})
	if len(seen) != 1 {
		t.Fatalf("calls = %v, want 1 same-cell pair only", seen)
	}
}
